flow-server/src/main/java/com/vaadin/flow/server/VaadinSession.java

```
@@ -544,6 +547,7 @@ public boolean hasLock() {
      *            and the session is not locked
      */
     public void checkHasLock(String message) {
+        assert deserializedAsEmpty == false : "The session was discarded during serialization and should not be used";
```

I think it should not be an `assert`.
`assert` is for Flow developers: it most likely will be disabled for the app developer (since you have to enable it explicitely via `-ea` ).
it should throw an exception.

---

Yes and this can only fail because of a bug in Flow. The error message has no meaning for anybody else

---

May be you are right.